### PR TITLE
Implement stream compression - transport layer

### PR DESCRIPTION
**Pull request by muxi**
The second portion of stream compression efforts (1st: #11189).

Modifications to chttp2 transport layer so that when stream compression is enabled (in the next PR), the messages can be compressed or decompressed with the stream compression lib module.

P.S. the first few commits (until f9ff554) are cherry-picked from #11189 as the changes depend on them.

### CODE REVIEW

**Comment by muxi** [comment]
test this please

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 1]
```diff
@@ -1658,10 +1667,42 @@ void grpc_chttp2_maybe_complete_recv_message(grpc_exec_ctx *exec_ctx,
...
+          GPR_ASSERT(s->decompressed_data_buffer.length == 0);
+          bool end_of_context;
+          if (!s->stream_decompression_ctx) {
+            s->stream_decompression_ctx =
+                grpc_stream_compression_context_create(
+                    GRPC_STREAM_COMPRESSION_DECOMPRESS);
+          }
+          if (!grpc_stream_decompress(s->stream_decompression_ctx,
+                                      &s->unprocessed_incoming_frames_buffer,
+                                      &s->decompressed_data_buffer, NULL, 5,
```
Comment (dgquintas) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: 5 is a magic number. Can it be made a `#define` with a descriptive name?

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 2]
```diff
@@ -1699,7 +1740,37 @@ void grpc_chttp2_maybe_complete_recv_trailing_metadata(grpc_exec_ctx *exec_ctx,
...
+       * maybe decompress the next 5 bytes in the stream. */
+      bool end_of_context;
+      if (!s->stream_decompression_ctx) {
+        s->stream_decompression_ctx = grpc_stream_compression_context_create(
+            GRPC_STREAM_COMPRESSION_DECOMPRESS);
+      }
+      if (!grpc_stream_decompress(s->stream_decompression_ctx,
+                                  &s->frame_storage,
+                                  &s->unprocessed_incoming_frames_buffer, NULL,
+                                  5, &end_of_context)) {
```
Comment (dgquintas) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: magic number 5

File: src/core/ext/transport/chttp2/transport/chttp2_transport.c [hunk 3]
```diff
@@ -2628,17 +2700,41 @@ static grpc_error *incoming_byte_stream_pull(grpc_exec_ctx *exec_ctx,
...
+        !s->unprocessed_incoming_frames_decompressed) {
+      bool end_of_context;
+      if (!s->stream_decompression_ctx) {
+        s->stream_decompression_ctx = grpc_stream_compression_context_create(
+            GRPC_STREAM_COMPRESSION_DECOMPRESS);
+      }
+      if (!grpc_stream_decompress(s->stream_decompression_ctx,
+                                  &s->unprocessed_incoming_frames_buffer,
+                                  &s->decompressed_data_buffer, NULL,
+                                  ~(size_t)0, &end_of_context)) {
```
Comment (dgquintas) [inline src/core/ext/transport/chttp2/transport/chttp2_transport.c]: put `~(size_t)0` into a const variable (or #define) with a descriptive name (eg SIZE_T_MAX)

File: src/core/ext/transport/chttp2/transport/internal.h [hunk 4]
```diff
@@ -518,6 +519,26 @@ struct grpc_chttp2_stream {
   grpc_chttp2_write_cb *on_write_finished_cbs;
   grpc_chttp2_write_cb *finish_after_write;
   size_t sending_bytes;
+
+  /** Whether stream compression send is enabled or not */
```
Comment (dgquintas) [inline src/core/ext/transport/chttp2/transport/internal.h]: nit: remove redundant "or not". Same in another comment right below.

**Comment by muxi** [comment]
Test failures:
Basic Tests Windows - #11745 
gRPC_interop_pull_requests - #10390 #10686

**Comment by ctiller** [comment]
This change causes a lot of duplication in critical code paths within chttp2.
I'd like to see another pass done to eliminate that.

One way to do so: have an identity stream compressor/decompressor algorithm, such that all bytes pass through it. This would eliminate needing the enabled bits, and all parsing and encoding duplication.

**Comment by muxi** [comment]
Agreed. Wilco.

**Pull request merged**